File: 3/C++/main.cpp
/**
* @author Matias Grioni
* @date 23:49 19/3/15
* @desc Third Project Euler problem. Finds largest prime
        factor of inputted number.
*/

#include <iostream>

using namespace std;

unsigned long largestPrimeFactor(unsigned long num) {
    unsigned long largest = 1;
    unsigned long limit = num;

    unsigned long i = 1;
    while(i < limit) {
        if(num % i == 0) {
            // If the largest prime factor of some number is 1, it is prime.
            // Woops no just got lucky, that's if the largest factor overall is.
            // The number can still have composite factors other than 1.
            if(largestPrimeFactor(i) == 1) {
                largest = i;
                limit = num / largest;
            }
        }

        i++;
    }

    return largest;
}

int main() {
    unsigned long input;
    cout << "Find largest prime factor > ";
    cin >> input;

    unsigned long factor = largestPrimeFactor(input);
    cout << endl << "Answer is: " << factor;

    return 0;
}

File: 3/Haskell/main.hs
{-
 - Author: Matias Grioni
 - Created: 7/7/15
 -
 - Finds the largest prime factor of a given number
 -}

module Main where

prime :: Integer -> Bool
prime n = null [x | x <- [2..floor . sqrt . fromIntegral $ n], n `mod` x == 0]

factors :: Integer -> [Integer]
factors n = [x | x <- reverse [2..floor . sqrt . fromIntegral $ n], n `mod` x == 0]

largestPrimeFactor :: Integer -> Integer
largestPrimeFactor n = head $ filter prime $ factors n

main = do
    putStrLn "Finds the largest prime factor of n"
    putStrLn "Enter n > "
    input <- getLine
    let n = read input :: Integer

    print $ largestPrimeFactor n
